File: lec10/02_tst/start/tst.c
// tst.c - 시작 코드
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

/**
 * TST (Ternary Search Tree) 데이터 구조 구현 (C 언어) - 시작 코드
 *
 * 시간 복잡도:
 * - 평균 삽입: O(log n)
 * - 평균 검색: O(log n)
 * - 평균 삭제: O(log n)
 * - 최악의 경우: O(n) (불균형 트리)
 * 공간 복잡도: O(n) where n is the number of nodes
 */

// TST 노드 구조
typedef struct TSTNode {
    char data;              // 현재 노드의 문자
    bool hasValue;          // 값 존재 여부
    int value;              // 저장할 값
    struct TSTNode* left;   // 현재 문자보다 작은 문자
    struct TSTNode* middle; // 다음 문자로 이동
    struct TSTNode* right;  // 현재 문자보다 큰 문자
} TSTNode;

// 결과를 저장하기 위한 구조체
typedef struct WordValuePair {
    char word[100];
    int value;
} WordValuePair;

// 함수 선언
TSTNode* createTSTNode(char data);
TSTNode* insert(TSTNode* root, const char* word, int value, int index);
TSTNode* insertWord(TSTNode* root, const char* word, int value);
int search(TSTNode* root, const char* word, int index);
int searchWord(TSTNode* root, const char* word);
bool startsWith(TSTNode* root, const char* prefix, int index);
bool startsWithPrefix(TSTNode* root, const char* prefix);
void getAllWords(TSTNode* root, WordValuePair* results, int* count);
void getWordsWithPrefix(TSTNode* root, const char* prefix, WordValuePair* results, int* count);
TSTNode* deleteWord(TSTNode* root, const char* word);
void destroyTST(TSTNode* root);

/**
 * 새로운 TST 노드 생성 및 초기화
 *
 * @param data 노드에 저장할 문자
 * @return 새로 생성된 TSTNode 포인터
 */
TSTNode* createTSTNode(char data) {
    // TODO: 새로운 TSTNode 생성 및 초기화
    // 1. 메모리 할당
    // 2. data 설정
    // 3. hasValue를 false로, value를 0으로 초기화
    // 4. left, middle, right를 모두 NULL로 초기화
    return NULL;
}

/**
 * TST에 단어와 값을 삽입 (재귀적 구현)
 *
 * @param root 현재 노드
 * @param word 삽입할 단어
 * @param value 저장할 값
 * @param index 현재 문자 인덱스
 * @return 루트 노드 포인터
 */
TSTNode* insert(TSTNode* root, const char* word, int value, int index) {
    // TODO: TST에 재귀적으로 단어 삽입
    // 1. 단어 끝 확인
    // 2. 현재 문자 가져오기
    // 3. 루트가 NULL이면 새 노드 생성
    // 4. 문자 비교하여 left/middle/right로 이동
    // 5. 마지막 문자일 때 값 저장
    return NULL;
}

/**
 * TST에 단어와 값을 삽입하는 래퍼 함수
 *
 * @param root TST의 루트 노드
 * @param word 삽입할 단어
 * @param value 저장할 값
 * @return 루트 노드 포인터
 */
TSTNode* insertWord(TSTNode* root, const char* word, int value) {
    // TODO: 입력 검증 후 insert 함수 호출
    return NULL;
}

/**
 * TST에서 단어 검색 (재귀적 구현)
 *
 * @param root 현재 노드
 * @param word 검색할 단어
 * @param index 현재 문자 인덱스
 * @return 값이 있으면 해당 값, 없으면 -1
 */
int search(TSTNode* root, const char* word, int index) {
    // TODO: TST에서 재귀적으로 단어 검색
    // 1. NULL 체크 및 인덱스 범위 확인
    // 2. 현재 문자 가져오기
    // 3. 문자 비교하여 left/middle/right로 이동
    // 4. 마지막 문자일 때 값 반환
    return -1;
}

/**
 * TST에서 단어 검색하는 래퍼 함수
 *
 * @param root TST의 루트 노드
 * @param word 검색할 단어
 * @return 값이 있으면 해당 값, 없으면 -1
 */
int searchWord(TSTNode* root, const char* word) {
    // TODO: 입력 검증 후 search 함수 호출
    return -1;
}

/**
 * 주어진 prefix로 시작하는 단어들이 있는지 확인 (재귀적 구현)
 *
 * @param root 현재 노드
 * @param prefix 검색할 접두사
 * @param index 현재 문자 인덱스
 * @return 존재하면 true, 없으면 false
 */
bool startsWith(TSTNode* root, const char* prefix, int index) {
    // TODO: 접두사 존재 여부 확인
    // 1. NULL 체크
    // 2. prefix 끝까지 도달했으면 true
    // 3. 문자 비교하여 재귀 호출
    return false;
}

/**
 * 접두사 검색 래퍼 함수
 *
 * @param root TST의 루트 노드
 * @param prefix 검색할 접두사
 * @return 존재하면 true, 없으면 false
 */
bool startsWithPrefix(TSTNode* root, const char* prefix) {
    // TODO: 입력 검증 후 startsWith 함수 호출
    return false;
}

/**
 * TST의 모든 단어와 값을 반환
 *
 * @param root TST의 루트 노드
 * @param results 결과를 저장할 배열
 * @param count 결과 개수를 저장할 변수
 */
void getAllWords(TSTNode* root, WordValuePair* results, int* count) {
    // TODO: 모든 단어 수집
    // 헬퍼 함수 사용 권장
    *count = 0;
}

/**
 * 주어진 접두사로 시작하는 모든 단어와 값을 반환
 *
 * @param root TST의 루트 노드
 * @param prefix 검색할 접두사
 * @param results 결과를 저장할 배열
 * @param count 결과 개수를 저장할 변수
 */
void getWordsWithPrefix(TSTNode* root, const char* prefix, WordValuePair* results, int* count) {
    // TODO: 접두사로 시작하는 모든 단어 수집
    // 1. prefix에 해당하는 노드 찾기
    // 2. 해당 노드 이하 모든 단어 수집
    *count = 0;
}

/**
 * TST에서 단어 삭제
 *
 * @param root TST의 루트 노드
 * @param word 삭제할 단어
 * @return 삭제 후 루트 노드 포인터
 */
TSTNode* deleteWord(TSTNode* root, const char* word) {
    // TODO: 단어 삭제 (재귀적 구현 권장)
    return root;
}

/**
 * TST의 모든 노드를 메모리에서 해제
 *
 * @param root 해제할 노드
 */
void destroyTST(TSTNode* root) {
    // TODO: 후위 순회로 모든 노드 해제
    // 1. NULL 체크
    // 2. 모든 자식 노드 먼저 해제 (left, middle, right)
    // 3. 현재 노드 해제
}

/**
 * 테스트 함수
 */
void testTST() {
    printf("=== TST (Ternary Search Tree) 테스트 ===\n\n");
    
    // TST 생성
    TSTNode* root = NULL;
    
    // 테스트 데이터 삽입
    printf("=== 단어 삽입 테스트 ===\n");
    struct {
        const char* word;
        int value;
    } testWords[] = {
        {"apple", 10}, {"app", 5}, {"application", 20},
        {"apply", 8}, {"banana", 15}, {"band", 12}, {"bandana", 18},
        {"cat", 25}, {"car", 30}, {"card", 35}
    };
    
    int numTestWords = sizeof(testWords) / sizeof(testWords[0]);
    
    for (int i = 0; i < numTestWords; i++) {
        root = insertWord(root, testWords[i].word, testWords[i].value);
        printf("삽입: %s (값: %d)\n", testWords[i].word, testWords[i].value);
    }
    printf("\n");
    
    // 검색 테스트
    printf("=== 검색 테스트 ===\n");
    const char* searchWords[] = {"apple", "app", "application", "appl", "banana", "dog", "car"};
    int numSearchWords = sizeof(searchWords) / sizeof(searchWords[0]);
    
    for (int i = 0; i < numSearchWords; i++) {
        int result = searchWord(root, searchWords[i]);
        if (result != -1) {
            printf("검색: %s -> 찾음, 값: %d\n", searchWords[i], result);
        } else {
            printf("검색: %s -> 없음\n", searchWords[i]);
        }
    }
    printf("\n");
    
    // 접두사 검색 테스트
    printf("=== 접두사 검색 테스트 ===\n");
    const char* prefixes[] = {"app", "ban", "ca", "xyz"};
    int numPrefixes = sizeof(prefixes) / sizeof(prefixes[0]);
    
    for (int i = 0; i < numPrefixes; i++) {
        bool exists = startsWithPrefix(root, prefixes[i]);
        printf("접두사: %s -> %s\n", prefixes[i], exists ? "존재함" : "존재하지 않음");
        
        if (exists) {
            WordValuePair results[100];
            int count = 0;
            getWordsWithPrefix(root, prefixes[i], results, &count);
            printf("  관련 단어들: ");
            for (int j = 0; j < count; j++) {
                printf("%s(%d)", results[j].word, results[j].value);
                if (j < count - 1) printf(", ");
            }
            printf("\n");
        }
    }
    printf("\n");
    
    // 삭제 테스트
    printf("=== 삭제 테스트 ===\n");
    const char* deleteWords[] = {"app", "banana", "card"};
    int numDeleteWords = sizeof(deleteWords) / sizeof(deleteWords[0]);
    
    for (int i = 0; i < numDeleteWords; i++) {
        printf("삭제 전 검색: %s -> %d\n", deleteWords[i], searchWord(root, deleteWords[i]));
        root = deleteWord(root, deleteWords[i]);
        printf("삭제 후 검색: %s -> %d\n", deleteWords[i], searchWord(root, deleteWords[i]));
        printf("\n");
    }
    
    // 전체 단어 출력
    printf("=== 남은 모든 단어 출력 ===\n");
    WordValuePair allResults[100];
    int allCount = 0;
    getAllWords(root, allResults, &allCount);
    
    for (int i = 0; i < allCount; i++) {
        printf("%s: %d\n", allResults[i].word, allResults[i].value);
    }
    printf("\n");
    
    // 메모리 해제
    destroyTST(root);
    
    printf("테스트 완료\n");
}

// 메인 함수
int main() {
    testTST();
    return 0;
}

File: lec10/02_tst/start/tst.py
# tst.py - 시작 코드

class TSTNode:
    """
    TST (Ternary Search Tree) 노드 클래스
    
    TST 노드 구조:
    - data: 현재 노드의 문자
    - has_value: 값 존재 여부
    - value: 저장할 값
    - left: 현재 문자보다 작은 문자
    - middle: 다음 문자로 이동
    - right: 현재 문자보다 큰 문자
    """
    
    def __init__(self, data):
        # TODO: TSTNode 초기화
        # 1. data 설정
        # 2. has_value를 False로 초기화
        # 3. value를 0으로 초기화
        # 4. left, middle, right를 None으로 초기화
        pass


class TST:
    """
    TST (Ternary Search Tree) 데이터 구조 구현
    
    시간 복잡도:
    - 평균 삽입: O(log n)
    - 평균 검색: O(log n)
    - 평균 삭제: O(log n)
    - 최악의 경우: O(n) (불균형 트리)
    공간 복잡도: O(n) where n is the number of nodes
    """
    
    def __init__(self):
        """TST 초기화"""
        # TODO: 루트 노드를 None으로 초기화
        pass
    
    def insert(self, word, value):
        """
        TST에 단어와 값을 삽입
        
        Args:
            word (str): 삽입할 단어
            value (int): 저장할 값
            
        Returns:
            bool: 성공하면 True, 실패하면 False
        """
        # TODO: 단어와 값 삽입
        # 1. 입력 검증
        # 2. 헬퍼 함수 호출 (_insert)
        # 헬퍼 함수는 재귀적으로 구현
        return False
    
    def search(self, word):
        """
        TST에서 단어 검색
        
        Args:
            word (str): 검색할 단어
            
        Returns:
            int: 값이 있으면 해당 값, 없으면 -1
        """
        # TODO: 단어 검색
        # 1. 입력 검증
        # 2. 헬퍼 함수 호출 (_search)
        # 헬퍼 함수는 재귀적으로 구현
        return -1
    
    def starts_with(self, prefix):
        """
        주어진 prefix로 시작하는 단어들이 있는지 확인
        
        Args:
            prefix (str): 검색할 접두사
            
        Returns:
            bool: 존재하면 True, 없으면 False
        """
        # TODO: 접두사 검색
        # 1. 입력 검증
        # 2. 헬퍼 함수 호출 (_starts_with)
        return False
    
    def delete(self, word):
        """
        TST에서 단어 삭제
        
        Args:
            word (str): 삭제할 단어
            
        Returns:
            bool: 성공하면 True, 실패하면 False
        """
        # TODO: 단어 삭제
        # 헬퍼 함수 사용 권장 (재귀적 구현)
        return False
    
    def get_all_words(self):
        """
        TST의 모든 단어와 값을 반환
        
        Returns:
            list: (단어, 값) 튜플의 리스트
        """
        # TODO: 모든 단어 수집
        # 헬퍼 함수 사용 권장
        return []
    
    def get_words_with_prefix(self, prefix):
        """
        주어진 접두사로 시작하는 모든 단어와 값을 반환
        
        Args:
            prefix (str): 검색할 접두사
            
        Returns:
            list: (단어, 값) 튜플의 리스트
        """
        # TODO: 접두사로 시작하는 모든 단어 수집
        # 1. prefix에 해당하는 노드 찾기
        # 2. 해당 노드 이하의 모든 단어 수집
        return []
    
    def get_sorted_words(self):
        """
        사전순으로 정렬된 모든 단어를 반환
        
        Returns:
            list: (단어, 값) 튜플의 리스트 (사전순)
        """
        # TODO: 사전순 정렬된 단어 반환
        # 중위 순회 사용
        return []
    
    def print_tst(self):
        """TST 구조를 시각적으로 출력 (선택사항)"""
        # TODO: TST 구조 출력 (선택사항)
        pass
    
    def size(self):
        """
        TST에 저장된 단어의 개수 반환
        
        Returns:
            int: 저장된 단어의 개수
        """
        # TODO: 저장된 단어 개수 반환
        return 0
    
    def is_empty(self):
        """
        TST가 비어있는지 확인
        
        Returns:
            bool: 비어있으면 True, 아니면 False
        """
        # TODO: 비어있는지 확인
        return True


def test_tst():
    """테스트 함수"""
    print("=== TST (Ternary Search Tree) 테스트 ===\n")
    
    # TST 생성
    tst = TST()
    
    # 테스트 데이터 삽입
    print("=== 단어 삽입 테스트 ===")
    test_words = [
        ("apple", 10), ("app", 5), ("application", 20),
        ("apply", 8), ("banana", 15), ("band", 12), ("bandana", 18),
        ("cat", 25), ("car", 30), ("card", 35)
    ]
    
    for word, value in test_words:
        success = tst.insert(word, value)
        print(f"삽입: {word} (값: {value}) - {'성공' if success else '실패'}")
    print()
    
    # 검색 테스트
    print("=== 검색 테스트 ===")
    search_words = ["apple", "app", "application", "appl", "banana", "dog", "car"]
    
    for word in search_words:
        result = tst.search(word)
        if result != -1:
            print(f"검색: {word} -> 찾음, 값: {result}")
        else:
            print(f"검색: {word} -> 없음")
    print()
    
    # 접두사 검색 테스트
    print("=== 접두사 검색 테스트 ===")
    prefixes = ["app", "ban", "ca", "xyz"]
    
    for prefix in prefixes:
        exists = tst.starts_with(prefix)
        print(f"접두사: {prefix} -> {'존재함' if exists else '존재하지 않음'}")
        
        if exists:
            words = tst.get_words_with_prefix(prefix)
            word_strs = [f"{word}({value})" for word, value in words]
            print(f"  관련 단어들: {', '.join(word_strs)}")
    print()
    
    # 사전순 정렬된 단어 출력
    print("=== 사전순 정렬된 모든 단어 ===")
    sorted_words = tst.get_sorted_words()
    
    for word, value in sorted_words:
        print(f"{word}: {value}")
    print()
    
    # 삭제 테스트
    print("=== 삭제 테스트 ===")
    delete_words = ["app", "banana", "card"]
    
    for word in delete_words:
        print(f"삭제 전 검색: {word} -> {tst.search(word)}")
        tst.delete(word)
        print(f"삭제 후 검색: {word} -> {tst.search(word)}")
        print()
    
    # 삭제 후 남은 단어 출력
    print("=== 삭제 후 남은 모든 단어 ===")
    remaining_words = tst.get_all_words()
    
    for word, value in remaining_words:
        print(f"{word}: {value}")
    print()
    
    print(f"총 저장된 단어 수: {tst.size()}")
    print(f"TST가 비어있는가: {tst.is_empty()}")
    
    print("\n테스트 완료")


if __name__ == "__main__":
    test_tst()